app/controllers/admin/worldwide_organisations_controller.rb

```diff
@@ -28,7 +30,12 @@ def update
     respond_with :admin, @worldwide_organisation
   end
 
-  def show; end
+  def show
+    @versions = Kaminari
+                  .paginate_array(@worldwide_organisation.versions_desc)
+                  .page(params[:page])
+                  .per(VERSIONS_PER_PAGE)
```

Comment:
We should avoid using `Kaminari.paginate_array` if at all possible.

That method is designed to handle ['dumb' Array objects](https://github.com/kaminari/kaminari#paginating-a-generic-array-object) which hold an entire result set and need to be sliced up in-memory. It has caused significant performance issues in the past on Documents with large edit histories. That's why we [stopped using it in the Editions controller](https://github.com/alphagov/whitehall/pull/6706/commits/5bd53cb37cc4812949d4b55553dcb8cbbe00e832) when we needed to [improve page load performance](https://trello.com/c/SKFiAakd/298-handle-large-change-histories-more-efficiently-in-whitehall).

In practice, `paginate_array` will load _every_ Version from the database into memory, and then immediately discard those which aren't needed for the current page.

Instead, we should be able to rely on Kaminari's ability to [paginate ActiveRecord results](https://github.com/kaminari/kaminari#query-basics) at the database layer – i.e. by using SQL's LIMIT and OFFSET clauses.

**This is untested**, but I expect the following will work:

```suggestion
    @versions = @worldwide_organisation.versions_desc
                  .page(params[:page])
                  .per(VERSIONS_PER_PAGE)
```

Reply:
Thanks for this. This was a TODO as I couldn't remember how to do it, but what you have wrote there works great! ⭐